api/src/test/java/org/pmiops/workbench/api/AppControllerTest.java

```diff
@@ -0,0 +1,118 @@
+import org.pmiops.workbench.exceptions.ForbiddenException;
+import org.pmiops.workbench.leonardo.ApiException;
+import org.pmiops.workbench.leonardo.LeonardoApiClient;
+import org.pmiops.workbench.leonardo.LeonardoApiHelper;
+import org.pmiops.workbench.model.App;
+import org.pmiops.workbench.model.AppType;
+import org.pmiops.workbench.model.WorkspaceAccessLevel;
+import org.pmiops.workbench.workspaces.WorkspaceAuthService;
+import org.pmiops.workbench.workspaces.WorkspaceService;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.config.ConfigurableBeanFactory;
+import org.springframework.boot.test.autoconfigure.orm.jpa.DataJpaTest;
+import org.springframework.boot.test.context.TestConfiguration;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Import;
+import org.springframework.context.annotation.Scope;
+
+@DataJpaTest
+public class AppControllerTest {
+  @TestConfiguration
+  @Import({
+    AppController.class,
+    FakeClockConfiguration.class,
+    LeonardoApiHelper.class,
+  })
+  static class Configuration {
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    public WorkbenchConfig workbenchConfig() {
+      return config;
+    }
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    DbUser user() {
+      return user;
+    }
+  }
+
+  @Autowired private AppController controller;
+
+  @MockBean LeonardoApiClient mockLeonardoApiClient;
+  @MockBean WorkspaceAuthService mockWorkspaceAuthService;
+  @MockBean WorkspaceService mockWorkspaceService;
+
+  private static final String GOOGLE_PROJECT_ID = "aou-gcp-id";
+  private static final String WORKSPACE_NS = "workspace-ns";
+  private static final String WORKSPACE_NAME = "workspace name";
+  private static final String WORKSPACE_ID = "myfirstworkspace";
+
+  private static WorkbenchConfig config = new WorkbenchConfig();
+  private static DbUser user = new DbUser();
+  private App testApp;
+  private DbWorkspace testWorkspace;
+
+  @BeforeEach
+  public void setUp() {
+    config = WorkbenchConfig.createEmptyConfig();
+    config.featureFlags.enableGkeApp = true;
+
+    testApp = new App().appType(AppType.RSTUDIO).googleProject(GOOGLE_PROJECT_ID);
+    testWorkspace =
+        new DbWorkspace()
+            .setWorkspaceNamespace(WORKSPACE_NS)
+            .setGoogleProject(GOOGLE_PROJECT_ID)
+            .setName(WORKSPACE_NAME)
+            .setFirecloudName(WORKSPACE_ID);
+    doReturn(testWorkspace).when(mockWorkspaceService).lookupWorkspaceByNamespace((WORKSPACE_NS));
+  }
+
+  @Test
+  public void testCreateAppSuccess() throws Exception {
+    controller.createApp(WORKSPACE_NS, testApp);
+    verify(mockLeonardoApiClient).createApp(testApp, WORKSPACE_NS, WORKSPACE_ID);
+  }
+
+  @Test
+  public void testCreateAppFail_featureNotEnabled() throws Exception {
```

Comment:
There's a few other cases (such as validating active billing) I see in the controller that could be covered here.

Reply:
added